### PR TITLE
Experiment listening to SDK eventsource logs from unit test

### PR DESCRIPTION
**Pull request by cijothomas**
## Changes
Adds a Testlistener for EventSource events from SDK.
This can be handy when investigating test failures related to threading issues etc. 

This is an initial attempt to see it in action. This PR tries to improve the test stability of one test by leveraging the above.

Please provide a brief description of the changes here.

For significant contributions please make sure you have completed the following items:

* [ ] Design discussion issue #
* [ ] Changes in public API reviewed

### CODE REVIEW

File: test/OpenTelemetry.Tests/Implementation/Trace/Export/BatchingActivityProcessorTests.cs [hunk 1]
```diff
@@ -274,21 +277,36 @@ public async Task ShutdownOnNotEmptyQueueFullFlush()
...
                                         .AddProcessorPipeline(pp => pp.AddProcessor(ap => activityProcessor)));
 
-            var activities = new List<Activity>();
-            for (int i = 0; i < 100; i++)
+            using (var testListener = new TestEventListener())
             {
-                activities.Add(this.CreateActivity(i.ToString()));
-            }
+                // Enable events from OpenTelemetrySDK.
+                testListener.EnableEvents(OpenTelemetrySdkEventSource.Log, EventLevel.Verbose);
```
Comment (cijothomas) [inline test/OpenTelemetry.Tests/Implementation/Trace/Export/BatchingActivityProcessorTests.cs:283]: move this to constructor?

File: test/OpenTelemetry.Tests/Implementation/Trace/Export/BatchingActivityProcessorTests.cs [hunk 2]
```diff
@@ -274,21 +277,36 @@ public async Task ShutdownOnNotEmptyQueueFullFlush()
...
+                using (var cts = new CancellationTokenSource(DefaultTimeout))
+                {
+                    await activityProcessor.ShutdownAsync(cts.Token);
+                }
 
-            Assert.True(activityExporter.WasShutDown);
-            Assert.Equal(activities.Count, activityExporter.ExportedActivities.Length);
-            Assert.InRange(exportCalledCount, activities.Count / batchSize, activities.Count);
+                // Get the shutdown event.
+                var shutdownEvent = testListener.Events.Where((e) => e.EventId == 2).First();
```
Comment (cijothomas) [inline test/OpenTelemetry.Tests/Implementation/Trace/Export/BatchingActivityProcessorTests.cs:297]: "2" is not magic. Its the eventid for the shutdownevent used by the BatchingProcessor.

File: test/OpenTelemetry.Tests/TestEventListener.cs [hunk 3]
```diff
@@ -0,0 +1,37 @@
...
+using System.Collections.Generic;
+using System.Diagnostics.Tracing;
+using System.Text;
+
+namespace OpenTelemetry.Tests
+{
+    /// <summary>
+    /// Uses in unit tests to listen to the SDK EventSource events.
+    /// </summary>
+    public class TestEventListener : EventListener
```
Comment (reyang) [inline test/OpenTelemetry.Tests/TestEventListener.cs:28]: Consider `InMemoryEventListener`?

File: test/OpenTelemetry.Tests/Implementation/Trace/Export/BatchingActivityProcessorTests.cs (hunk 2 above)
Comment (reyang) [reply test/OpenTelemetry.Tests/Implementation/Trace/Export/BatchingActivityProcessorTests.cs:297]: Might need to make this an internal enum and make it visible to test project.

File: test/OpenTelemetry.Tests/Implementation/Trace/Export/BatchingActivityProcessorTests.cs [hunk 4]
```diff
@@ -274,21 +277,36 @@ public async Task ShutdownOnNotEmptyQueueFullFlush()
                                         .SetSampler(new AlwaysOnActivitySampler())
                                         .AddProcessorPipeline(pp => pp.AddProcessor(ap => activityProcessor)));
 
-            var activities = new List<Activity>();
-            for (int i = 0; i < 100; i++)
+            using (var testListener = new TestEventListener())
```
Comment (reyang) [inline test/OpenTelemetry.Tests/Implementation/Trace/Export/BatchingActivityProcessorTests.cs:280]: Probably a minor improvement of readability and save some indentation spaces:

```suggestion
            using var testListener = new TestEventListener();
```

File: test/OpenTelemetry.Tests/Implementation/Trace/Export/BatchingActivityProcessorTests.cs [hunk 5]
```diff
@@ -286,8 +290,20 @@ public async Task ShutdownOnNotEmptyQueueFullFlush()
...
 
+            // Get the shutdown event.
+            // 2 is the EventId for OpenTelemetrySdkEventSource.ShutdownEvent
+            // TODO: Expose event ids as internal, so tests can access them more reliably.
+            var shutdownEvent = testListener.Events.Where((e) => e.EventId == 2).First();
+
+            int droppedCount = 0;
+            if (shutdownEvent != null)
+            {
+                // There is a single payload which is the number of items left in buffer a shutdown.
```
Comment (reyang) [inline test/OpenTelemetry.Tests/Implementation/Trace/Export/BatchingActivityProcessorTests.cs:301]:
```suggestion
                // There is a single payload which is the number of items left in buffer at shutdown.
```

File: test/OpenTelemetry.Tests/Implementation/Trace/Export/BatchingActivityProcessorTests.cs [hunk 6]
```diff
@@ -286,8 +290,20 @@ public async Task ShutdownOnNotEmptyQueueFullFlush()
                 await activityProcessor.ShutdownAsync(cts.Token);
             }
 
+            // Get the shutdown event.
+            // 22 is the EventId for OpenTelemetrySdkEventSource.ForceFlushCompleted
```
Comment (CodeBlanch) [inline test/OpenTelemetry.Tests/Implementation/Trace/Export/BatchingActivityProcessorTests.cs:294]: nit: Comment is for 22 but the code is using 2.

**Pull request merged**